cmd/link/link.go

```diff
@@ -0,0 +1,41 @@
+package link
+
+import (
+	"fmt"
+
+	"github.com/ncw/rclone/cmd"
+	"github.com/ncw/rclone/fs/operations"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	cmd.Root.AddCommand(commandDefintion)
+}
+
+var commandDefintion = &cobra.Command{
+	Use:   "link remote:path",
+	Short: `Generate public link to file/folder.`,
+	Long: `
+rclone link will create or retrieve a public link to the given file or folder.
+
+    rclone link remote:path/to/file
+    rclone link remote:path/to/folder/
+
+If successful, the last line of the output will contain the link. Exact
+capabilities depend on the remote, but the link will always be created with
+the least constraints – e.g. no expiry, no password protection, accessible
+without account.
+`,
+	Run: func(command *cobra.Command, args []string) {
+		cmd.CheckArgs(1, 1, command, args)
+		fsrc, srcFileName := cmd.NewFsSrcWithFile(args)
```

Comment:
I would have thought you'd be better calling [newFsFile](https://github.com/ncw/rclone/blob/master/cmd/cmd.go#L147) and then you'll know whether the argument points to a file or directory without the need for a trailing `/` check which I've tried to avoid in rclone.  (You'd have to make it public first though...)

Then the PublicLink is then either sharing the root directory (if passed "") or a path within that directory.

I don't know - what do you think?

Reply:
You are most certainly right, but I ran into issues previously and it was not clear to me where they were coming from. With Dropbox, I can simply drop in `NewFsFile` and it continues to work.

However, with Drive I run into the peculiar issue that the `f.root` *changes*. I added some debug output in the form of `fmt.Printf("%s @ %s  0\n", f.Root(), remote)` to the code. It's still fine in `operations.PublicLink`. It's fine at the top of Dropbox's `PublicLink`. However, looking at the debug output for Drive (`xxlink` is a file at the very root, i.e. no subroot):

```
$ go build -i && ./rclone -vvv link Gsuite:xxlink
2018/03/24 19:36:01 DEBUG : rclone: Version "v1.40-DEV" starting with parameters ["./rclone" "-vvv" "link" "Gsuite:xxlink"]
2018/03/24 19:36:01 DEBUG : Using config file from "/home/stefan/.rclone.conf"
 @ xxlink
 @ xxlink  0
xxlink @ xxlink  1
2018/03/24 19:36:02 DEBUG : Google drive root 'xxlink': attempting to share single file 'xxlink'
xxlink @ xxlink  2
drive fs: Google drive root 'xxlink'
2018/03/24 19:36:02 Failed to link: object not found
```

For comparison, this is the output for Dropbox (again `rclone139` is a top level file):

```
$ go build -i && ./rclone -vvv link dropbox:rclone139
2018/03/24 19:35:01 DEBUG : rclone: Version "v1.40-DEV" starting with parameters ["./rclone" "-vvv" "link" "dropbox:rclone139"]
2018/03/24 19:35:01 DEBUG : Using config file from "/home/stefan/.rclone.conf"
 @ rclone139
 @ rclone139  0
 @ rclone139  1
2018/03/24 19:35:02 DEBUG : Dropbox root '': attempting to share 'rclone139' (absolute path: /rclone139)
2018/03/24 19:35:02 DEBUG : /rclone139: has a public link already, attempting to retrieve it
https://www.dropbox.com/s/otsnp322yzdetup/rclone139?dl=0
2018/03/24 19:35:03 DEBUG : 4 go routines active
2018/03/24 19:35:03 DEBUG : rclone: Version "v1.40-DEV" finishing with parameters ["./rclone" "-vvv" "link" "dropbox:rclone139"]
```

So I guess there must be a bug somewhere in the drive code? Maybe `rootFolderId` or some of the dirCache is not cleaned properly… if you have a hunch, please let me know.

Reply:
Hm, it seems that the old or incorrect FS resurfaces somehow… I print the memory addresses in drive's `NewFs` implementation here: https://github.com/ncw/rclone/blob/master/backend/drive/drive.go#L571

```
in NewFs:
old: 0xc4203b4000 (f)
new: 0xc42057a090 (newF)

in operations.PublicLink:
0xc42057a090 (i.e. the correct one)

in drive.PublicLink:
0xc4203b4000 (i.e. the incorrect one)
```

so I guess `Features()` hangs on to the old reference somehow?

Reply:
So, it's exactly like that: the `f` features are filled with functions whose receiver is `*f`. These are copied when `newF` is created and are never updated – they keep `*f` as their receiver. I guess calling `.Fill()` again might work, but so does simply overwriting `f`'s `dirCache` and `root`. 

A test that ensures that the address of `f` matches the addresses of all receivers of `f.Features()` functions sounds like a decent idea. Unfortunately I could not find a way to get the pointer of a func's *receiver* from the func's own pointer.